test/Microsoft.Extensions.Internal.Test/ClosedGenericMatcherTest.cs

```diff
@@ -174,6 +174,65 @@ public class ClosedGenericMatcherTest
             Assert.Equal(expectedResult, result);
         }
 
+        // IEnumerable<int> is preferred because it is defined on the more-derived type.
+        [Fact]
+        public void ExtractGenericInterface_MultipleDefinitionsInherited()
+        {
+            // Arrange
+            var type = typeof(TwoIEnumerableImplementationsInherited);
+
+            // Act
+            var result = ClosedGenericMatcher.ExtractGenericInterface(type, typeof(IEnumerable<>));
+
+            // Sort
+            Assert.Equal(typeof(IEnumerable<int>), result);
+        }
+
+        // IEnumerable<int> is preferred because we sort by Ordinal on the full name.
+        [Fact]
+        public void ExtractGenericInterface_MultipleDefinitionsOnSameType()
+        {
+            // Arrange
+            var type = typeof(TwoIEnumerableImplementationsOnSameClass);
+
+            // Act
+            var result = ClosedGenericMatcher.ExtractGenericInterface(type, typeof(IEnumerable<>));
+
+            // Sort
+            Assert.Equal(typeof(IEnumerable<int>), result); 
+        }
+
+        private class TwoIEnumerableImplementationsOnSameClass : IEnumerable<string>, IEnumerable<int>
+        {
+            IEnumerator<int> IEnumerable<int>.GetEnumerator()
+            {
+                throw new NotImplementedException();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                throw new NotImplementedException();
+            }
+
+            IEnumerator<string> IEnumerable<string>.GetEnumerator()
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private class TwoIEnumerableImplementationsInherited : List<string>, IEnumerable<int>
```

Comment:
I would switch the `string` and `int` here because this could pass even without the "search more derived types" logic (`int` < `string` anyway).